### ISSUE TITLE
A hard HW4 idea that would get you lots of extra credit

### ISSUE THREAD

**Issue by mikeizbicki** [mikeizbicki/ucr-cs100#344]
There are two problems with how we're currently testing homework submissions using the `script` command:
1. Running tests manually is not "hacker-like".  In the real world, programs use automated testing solutions.  Unfortunately, there is no way to automate the `script` command.
2. When manually reading script files, it is difficult for the grader to figure out exactly what tests the student is trying to cover.

**A project that improves this situation can earn up to 200 points on this assignment.** This will be a relatively hard task requiring you to develop a solution on your own. The solution must:
1.  Fix one (or preferably both) of the pain points described above.
2.  Be simple enough that students can use it on their first homework submission with minimal explanation.

I think the most promising idea is to use a [unit testing framework](https://en.wikipedia.org/wiki/Unit_testing).  Unfortunately, there are no unit testing frameworks designed to test shell features, which is what we need.  You would have to look at existing frameworks, and figure out a good way to use them to test rshell.  I would recommend starting your research by reading [this stackoverflow question](http://stackoverflow.com/questions/1339416/unit-testing-bash-scripts) about unit testing shell scripts.

If you're interested in pursuing this, you would have to work pretty closely with me to ensure that your final product would be usable by future classes.

**Comment by gbrlrz017** [comment]
Just to see if I understand the issues correctly:

For 1, you want something that tests the student's program for different cases? For example, testing all the permutations and combinations possible for entering ls with the flags a, l, or R ( "ls -alR", "ls -aRl", "ls -laR", "ls -a -R -l", ... , "ls -a", "ls -l", "ls -R", "ls"), and anything else that could possibly break the program. Would this be for the students to use or the instructors? 

For 2, you want the script file students submit be more easily readable and/or for the instructor to receive a checklist of some sort saying what was tested?

**Comment by mikeizbicki** [comment]
@gbrlrz017 
1.  It would be for the students to use.  The students would just write a list of test cases, then some tool (the unit testing library) would go and actually run the test cases.  The grader would then be able to easily reproduce the test cases by running the same tool; they could also easily inspect the code that generated the tests without running the tests.
2. It doesn't necessarily have to be a script file anymore.  So something like the checklist idea would be ideal.

**Comment by KenleyArai** [comment]
@mikeizbicki 

Would it be appropriate to make a tutorial on unit testing in C++ instead of testing scripts? 

Since we are programming in C++ and not shell wouldn't it make more sense to write testing for our C++ functions i.e:

```
assert( trim_extra_white_space("  ls   -r     -a   "), string("ls -r -a")); 
assert( command("pwd #this command has a comment"), string("pwd"));
```

This wouldn't completely solve the issue of running the tests via script, and it will shift the responsibility to the student quite a bit more because of the additional libraries to learn. However it will be much clearer to the grader how much the programmer tested. 

I believe this will improve the overall quality of code because much like gdb or vim, unit testing helps programmers understand what is happening within their code.

If the above is appropriate for hw4 I would like to write up:
- How to write tests in C++
- A development style that uses testing, [TDD](http://en.wikipedia.org/wiki/Test-driven_development)

**Comment by mikeizbicki** [comment]
@KenleyArai This is on the right track.  `assert` statements like you have go a long ways towards convincing us that your program works correctly.  But they don't go all the way.  Somehow, we need to test that when the user enters the command

```
ls -la /home/user
```

that the `ls` program actually gets called with the parameters `-la` and `/home/user`.

One way to do this in C++ would be to have students write a "shell library" that provides a function:

```
int system(char *cmd);
```

The `cmd` string would be the command typed in (possibly with parameters, connectors, comments, and whatever else a user might decide to enter).  Then you could create test cases like:

```
register_test("ls");
register_test("ls -la");
register_test("ls -la /home/user");
register_test("ls a q w e r t y  u i  o p asd d f g h h j j k z x ");
```

where `register_test` has the following pseudocode:

```
register_test(char *cmd)
{
     call studentlibrary::system(cmd), capture stdout into "stdout.student"
     call bash::system(cmd), capture stdout into "stdout.bash"
     return stdout.student == stdout.bash
}
```

**tldr** A good unit testing tutorial is a great idea.  You'd have to do a little bit more work, though, to get the extra credit.  I'm perfectly okay with making students learn the unit testing framework, since any possible solution would require this.

Do you have a particular library in mind that you like?

**Comment by KenleyArai** [comment]
@mikeizbicki

I have a little experiace with cppunit, however I am much more interested in learning the newer library from google

To solve the issue of testing commands I believe it is doable using popen to capture the output from stdout.

I think a change that could be applied for future classes, would be to have students implement something similar to `/bin/bash -c 'ls'`. This change would enable easy captureing of output i.e:

```
bool register_test( cmd )
{
    from_bash = capture("/bin/bash -c" + cmd);
    from_rshell = capture("bin/rshell -c" + cmd);

    return from_bash == from_rshell;
}
```

If this change is **not** okay I do have ideas on how to accouplish the extra credit assignment without the `-c` command.

Either way I would like to take on this assignment.

Would you prefer me making the library for automatically testing, or writing a tutorial for unit testing, or both?

**Comment by mikeizbicki** [comment]
You shouldn't write something completely from scratch.  It should just be a very thin wrapper around some full featured unit testing library, basically like a template that students would then fill in.  I don't particularly care which library gets used.  Why do you like the google one?  

Doing the `-c` flag is an interesting idea.  Right now that sounds pretty good to me, but I'll have to think about it a bit more to make sure.  What were your other ideas?

Ultimately what I want is a detailed set of instructions for how students should be writing tests.  So writing up a tutorial along with some examples would be it.

**Comment by KenleyArai** [comment]
Boost  tends to be quite difficult for some to install, and I personally don't like cppunit. Google seems like a fair comprimise between the two. However as I said I only really know cppunit, and I'm just trying Google because it seems like a well maintained and straight forward library. My opinion may change over time. :)

I went ahead and made a proof of concept and tested it in my [rshell under unit_testing](https://github.com/KenleyArai/rshell/tree/unit_testing) branch.

I made a super primitive version of `rshell -c`.

In the `tests/test.cpp` it grabs the output from `bin/rshell -c 'ls -la'` and `/bin/bash -c 'ls -la'` output  and prints them out. 

Additionally if `-c` is not implemented it will also work from within rshell to grab the output, and test it against bash.

**Comment by jmoon018** [comment]
@mikeizbicki 

I made some progress with this program: https://github.com/jmoon018/rshell_grader

Currently, it can:
1) Store a list of commands (the file: /tests/input) that will be tested by the program
2) Create two directories for rshell and bash that can be used to make pwd/ls -l checks without affecting the student's environment. This will find out if cd/mkdir really works. However, this isn't completed yet
3) Creates instances of rshell and bash, inputs the commands from the input file, and stores the output
4) The output is compared to each other and stored as a pass or fail (along with the command name)
5) The results (list of commands + pass/fail) are printed before the program ends

The good thing is that it doesn't require any student intervention; none of the student's files have to change. Also surprisingly, the source code isn't that large--about ~250 lines, and ~30% of that is comments/whitespace

Ultimately, the autograder executable must be stored in the bin folder (and EXECUTED THERE TOO), and everything _should_ work

There may still be some features that need to be added before it becomes practical. So far, though, is this roughly what you would like to see in the program, Professor? Go ahead and try it out, any comments/critique/questions/etc are welcome

**Comment by hgarc014** [comment]
That sounds neat. some things that would be neat (if i were the grader) would be:
-  one file that only contains fail cases (would be text added to student git repo)
-  file that contains tested cases (i can see what is tested and add to it)
-  script to run on all students in class (may already exist) that way i don't have to manually execute the script for each student (maybe update student repos,all after all grades finished)

I did not completely read everything, but these are some ideas that would be nice and make grading easier. As for matching output of program with a correct solution:

This will likely cause problems. Especially for something like `ls` since there are multiple valid output solutions from the program

**Comment by jmoon018** [comment]
@hgarc014 
1. Creating a file with the fail cases seems like a good idea. It should be easy to implement, too
2. Currently, there is a text file called input which does exactly that --- it has a list of commands/test cases which are directly used by the program (delimited by new lines). You can delete or add test cases very easily
3. Yes, that is an important and massively time-saving feature. I can probably loop through the people/students/ folder, retrieve the github url for each student, clone and test, so everything gets done at once. I'll make sure to do this when I get the chance. The professor would love this.

In regards to the variations of the "ls" output, could you explain how that would happen? Are you referring to HW#1 (in which case, I can understand output differences and try to fix that), or HW#0 (if <i>execvp</i> invokes <i>ls</i>, shouldn't the output be the same)? I think demanding the exact same output will cause problems, but once I see some examples of valid but different outputs, I'll have a better understanding of when it occurs and how to fix it. For example, I've seen and know that error messages may be different between bash and rshell, so once bash outputs an error (not from an echo statement), and if the rshell line contains "error," I can mark it as passing

Thanks for the response and suggestions! The ideas you brought up are crucial to making this program viable for the Professor, and I will work on it when I get the chance.

**Comment by hgarc014** [comment]
well, I was referring to the HW#1 assignment. As for execvp calls I think the output would be the same in most cases (I can't currently think of one that would be different, I don't think it would be), other than errors as you pointed out. 

However, something like terminal output can lead to an error. For example, the rshell program can print out a "$" (if people did not do extra credit) with commands. It can also have "myUsr@machine" which will have variations on different user/machine combinations, but this will probably be fine considering it being run on the graders computer. That's about all I can think about for now.

**Comment by jmoon018** [comment]
I figured you were referring to HW#1---I just wanted to make sure. For the LS program, I can probably put each word/file into 2 lists--one for bash, the other for rshell---and compare them at the end. This way, any difference in how the files are ordered out will not matter. Additionally, this would remove any silly discrepancies between the rshell/bash (eg: one program uses tab instead of spaces). I'm not sure how to detect if color works. I think I'll be able to grasp the issue better after I start HW#1.._cough_

In regards to the innate passive output, I don't think it will be a big issue. If I recall correctly, the terminal output like $jamal-p15M: or whatever is ignored and not considered when comparing the bash/rshell files (HOWEVER I COULD BE WRONG --- I CAN'T CONFIRM ON MY LINUX MACHINE YET). Even if it isn't ignored, I can simply do some output editing --- I could delete the beginning portion of a line up until it states the command. So for example, if the command is "echo hi" and the terminal is $jamalsMachine: echo hi, I can go in automatically delete everything before the "echo hi" begins

Thanks! If anyone has more edge cases, ideas, or questions, let me know

**Comment by mikeizbicki** [comment]
@jmoon018 

I like what I see so far!  I haven't had a chance to look into this too closely yet though.  Can you come by office hours this week and do a quick demo for me so I can ask question as you go?  I have some more questions/comments, but I think it'll be easier to explain them in person.

Again, looks like some nice progress!

**Comment by jmoon018** [comment]
@mikeizbicki 
Sure, I'd be happy to. I'll continue to make some changes to it and produce a quick presentation before showing you. If there's anything specific you want to see, let me know, and I can probably implement it before Thursday

Thanks.
